scripts/coverage-to-json.py

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python
+
+import sys
+import fileinput
+import json
+import re
+
+if __name__ == "__main__":
+    filedict = {}
+    # Input looks like this:
+    #
+    #    27.16 %     85/313    src/avl.ml
+    #    96.85 %    246/254    src/burrow.ml
+    #   ...
+    #   100.00 %      4/4      src/ptr.ml
+    #    81.86 %   1859/2271   Project coverage
+    #
+    # So group(1) holds the percentage, group(2)/group(3) is the ratio, and
+    # group(4) captures the filename (or "Project coverage" in the last line)
+    p = re.compile("\s*(\d+\.\d+)\s*%\s*(\d+)/(\d+)\s*([a-zA-Z](.*))")
+
+    for line in fileinput.input():
+        match_object = re.search(p, line)
+        if match_object:
+            percentage = float(match_object.group(1))
+            filename = match_object.group(4)
+            prefix = "src/"
+            if filename.startswith(prefix):
+                # an ocaml file
+                filename = filename[len(prefix) :]
+            elif filename == "Project coverage":
+                # the summary for the entire project
+                filename = "TOTAL"
+            else:
+                sys.exit("malformed filename in coverage report: {}".format(filename))
+        else:
+            sys.exit("malformed line in coverage report: {}".format(line))
+
+        filedict[filename] = percentage
+
+    json_object = json.dumps(filedict, sort_keys=True, indent=4)
+    print(json_object)
```

Comment:
There's a trade-off of conciseness here, but we could inline this script to `Makefile` with something like:

```
bisect-ppx-report summary --per-file \
  | sed -E 's/^ *([0-9]+\.[0-9]+) *% *[^ ]* *(.*)$/\1|\2/g' \
  | jq -R 'split("|") | { "value": .[0] | tonumber, "key": (.[1] | if . == "Project coverage" then "TOTAL" else ltrimstr("src/") | rtrimstr(".ml") end) }' \
  | jq -s 'from_entries'
```

Arguably, once one knows a bit about the `jq` syntax this version might even be more readable.

Reply:
Haha, live and learn! Yeah, `jq` seems good for this kind of task, though the conditional in this case makes it a little less appealing to (the future) me (if I were less thorough with the error messages and the comments I think I could probably make the python code fit in 5 lines too :stuck_out_tongue_closed_eyes:). I'll use your `jq` script, since it saves us from having another file in `scripts/` :+1: